source/Mlos.Python/mlos/Optimizers/OptimizerBase.py

```diff
@@ -65,24 +78,97 @@ def optimum(self, stay_focused=False) -> Tuple[Point, Point]: # pylint: disable=
+        config_at_optimum = Point.from_dataframe(features_df.loc[[index_of_best]])
+        return config_at_optimum, optimum_value
+
+    @trace()
+    def _prediction_based_optimum(self, features_df: pd.DataFrame, optimum_definition: OptimumDefinition, alpha: float)-> Tuple[Point, Point]:
+        objective = self.optimization_problem.objectives[0]
+
+        # Let's see if we have them cached before recomputing
+        #
+        if self.cached_predictions_for_observations is None:
+            self.cached_predictions_for_observations = self.predict(feature_values_pandas_frame=features_df)
+        predictions_df = self.cached_predictions_for_observations.get_dataframe()
+
+        if len(predictions_df.index) == 0:
+            raise ValueError("Insufficient data to compute confidence-bound based optimum.")
+
+        # Predictions index must be a subset of features index.
+        #
+        assert features_df.index.intersection(predictions_df.index).equals(predictions_df.index)
+
+        predicted_value_column_name = Prediction.LegalColumnNames.PREDICTED_VALUE.value
+        dof_column_name = Prediction.LegalColumnNames.DEGREES_OF_FREEDOM.value
+        variance_column_name = Prediction.LegalColumnNames.PREDICTED_VALUE_VARIANCE.value
+
+        if optimum_definition == OptimumDefinition.PREDICTED_VALUE_FOR_OBSERVED_CONFIG:
+            if objective.minimize:
+                index_of_best = predictions_df[predicted_value_column_name].idxmin()
+            else:
+                index_of_best = predictions_df[predicted_value_column_name].idxmax()
+
+            optimum_value = Point.from_dataframe(predictions_df.loc[[index_of_best], [predicted_value_column_name]])
 
-        if self.optimization_problem.objectives[0].minimize:
-            index_of_best_target = objectives_df.idxmin()[0]
         else:
-            index_of_best_target = objectives_df.idxmax()[0]
-        best_objective = Point.from_dataframe(objectives_df.loc[[index_of_best_target]])
-        best_config_point = Point.from_dataframe(features_df.loc[[index_of_best_target]])
+            # We will be manipulating this data so let's make a copy for now.
+            # TODO: Profile this and if necessary optimize this copy away.
+            #
+            predictions_df = predictions_df.copy(deep=True)
+
+            # Drop nulls and zeroes.
+            #
+            predictions_df = predictions_df[predictions_df[dof_column_name].notna() & predictions_df[dof_column_name] != 0]
+
+            if len(predictions_df.index) == 0:
+                raise ValueError("Insufficient data to compute confidence-bound based optimum.")
+
+            t_values = scipy.stats.t.ppf(1 - alpha / 2, predictions_df[dof_column_name])
+            prediction_interval_radii = t_values * np.sqrt(predictions_df[variance_column_name])
+
+            if optimum_definition == OptimumDefinition.UPPER_CONFIDENCE_BOUND_FOR_OBSERVED_CONFIG:

```

Comment:
I'm a bit confused why this is nested in the
```python
if optimum_definition == OptimumDefinition.PREDICTED_VALUE_FOR_OBSERVED_CONFIG
```

Reply:
It's not, just the diff rendering makes the `else` a bit obscure. 
![image](https://user-images.githubusercontent.com/51930563/93914583-99583480-fcbb-11ea-9bbc-0c49c5b228b8.png)

Reply:
ohh that makes so much more sense, as I said, not the right amount of coffee.